[Accordion] Fix invalid HTML inside heading

ZeeshanTamboli opened the pull request:
Fixes #44153  

This PR resolves a regression introduced in #42914 by making the following changes to ensure valid HTML structure for the Accordion component:  
- Changed the Accordion summary element to `button`.
- Updated the Accordion summary content and icon wrapper elements to `span`.
- Keep `Typography` on the demos as before, but with `component='span'` since the `p` element rendered by Typography is invalid inside a `span`.

### Minor version change

- The HTML elements of the Accordion summary have been updated:
  - The root element is now `button` (previously `div`).  
  - Summary content and the icon wrapper are now `span` (previously `div`).  

How might this impact you:
- Developers using the previous `div` element for styling in the AccordionSummary should update their styling.

### How to fix similar invalid HTML in your codebase

Part of this PR is fixing invalid HTML in our demos. If you copied code from these demos in the past, you might have to do similar changes to your codebase.

If you are using `Typography` inside the `AccordionSummary`, which defaults to rendering a `p` tag,  you should use the `component` prop to replace the HTML tag as such: `<Typography component="span" />`, as shown in the [updated demos](https://deploy-preview-44408--material-ui.netlify.app/material-ui/react-accordion/).

---

These changes are critical to fix the regression but introduce a breaking change. IMO, this should be released in a minor version (`v6.x`) and documented in the migration guide. I know this is not ideal but the fix is also important.

---

If approved, I’ll include a note in the migration guide detailing these changes.

ZeeshanTamboli commented:
Any review even though it'a a breaking change?

aarongarciah commented on docs/data/material/components/masonry/MasonryWithVariableHeightItems.js:
```diff
@@ -27,7 +26,7 @@ export default function MasonryWithVariableHeightItems() {
             <StyledAccordion sx={{ minHeight: height }}>
               <AccordionSummary expandIcon={<ExpandMoreIcon />}>
-                <Typography>Accordion {index + 1}</Typography>
+                <span style={{ textAlign: 'left' }}>Accordion {index + 1}</span>
```
Why is `textAlign: left` needed?

ZeeshanTamboli replied:
The text alignment was left-aligned previously because `Typography` had that as its default, preventing Argos changes. 

### Left-aligned text: 

![image](https://github.com/user-attachments/assets/d098a6ca-4e13-4d69-b940-29a34fc82323)  

### Without left alignment:  

![Screenshot (30)](https://github.com/user-attachments/assets/134f1e76-79fa-498c-b04f-d0c1f3f45426)  

It's probably fine to leave it without left alignment.

aarongarciah replied:
What is causing the change in the alignment? The removal of `<Typography>` from the demos? If that's the case, I wonder if `AccordionSummary` should align text to the left (start) by default.

ZeeshanTamboli replied:
> What is causing the change in the alignment? The removal of <Typography> from the demos?

Yes, I also noticed this in the Accordion demos during screen resize. The issue is due to changing the summary content from a block element (`div`) to an inline element (`span`). The block-level text aligns to the left for LTR languages and inline elements like `span` do not control text alignment themselves. Adding `text-align: start` to the `<span>` fixed it. Nice catch!

aarongarciah commented on docs/data/material/components/accordion/ControlledAccordions.tsx:
```diff
@@ -21,10 +44,8 @@ export default function ControlledAccordions() {
-            General settings
-          </Typography>
-          <Typography sx={{ color: 'text.secondary' }}>I am an accordion</Typography>
+          <AccordionSummaryText>General settings</AccordionSummaryText>
```
Reading `AccordionSummaryText` makes me think this component is part of the `Accordion` API. Can we find another name that doesn't use the "Accordion" prefix?

ZeeshanTamboli replied:
Make sense. Changed it to `SummaryText`.

aarongarciah commented on docs/data/material/components/accordion/ControlledAccordions.js:
```diff
@@ -3,8 +3,29 @@ import Accordion from '@mui/material/Accordion';
 
+const SummaryText = styled('span')(({ theme }) => ({
+  margin: 0,
+  textAlign: 'left',
```
Is this needed now that `AccordionSummaryContent` has `textAlign: start`?

ZeeshanTamboli replied:
Not needed. Removed it.

aarongarciah commented on docs/data/material/components/accordion/ControlledAccordions.js:
```diff
@@ -3,8 +3,29 @@ import Accordion from '@mui/material/Accordion';
 import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
 
+const SummaryText = styled('span')(({ theme }) => ({
+  margin: 0,
```
Is this margin needed?

ZeeshanTamboli replied:
Not needed. Removed it.

aarongarciah commented on docs/data/material/components/accordion/ControlledAccordions.js:
```diff
@@ -3,8 +3,29 @@ import Accordion from '@mui/material/Accordion';
+import { styled } from '@mui/material/styles';
 import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
 
+const SummaryText = styled('span')(({ theme }) => ({
+  margin: 0,
+  textAlign: 'left',
+  variants: [
+    {
+      props: { secondary: true },
+      style: {
+        color: theme.palette.text.secondary,
+      },
+    },
+    {
+      props: ({ secondary }) => !secondary,
+      style: {
+        width: '33%',
+        flexShrink: 0,
+      },
+    },
+  ],
+}));
```
I think having two different components would make the demo easier to follow.

```suggestion
const SummaryLabel = styled('span')(({ theme }) => ({
  width: '33%',
  flexShrink: 0,
}));

const SummaryText = styled('span')(({ theme }) => ({
  color: theme.palette.text.secondary,
}));
```

ZeeshanTamboli replied:
Makes sense. Done.

aarongarciah commented:
@ZeeshanTamboli should there be any visual changes? I see one regression test changing due to typographic styles: https://app.argos-ci.com/mui/material-ui/builds/34147/123626742. I wonder if it should look the same given this PR is supposed to fix only the HTML structure and I wonder how many users will be affected by this change. Just asking in case we can avoid any visual change.

aarongarciah commented:
I just added the `v7.x`.

ZeeshanTamboli commented:
> @ZeeshanTamboli should there be any visual changes? I see one regression test changing due to typographic styles: https://app.argos-ci.com/mui/material-ui/builds/34147/123626742. I wonder if it should look the same given this PR is supposed to fix only the HTML structure and I wonder how many users will be affected by this change. Just asking in case we can avoid any visual change.
 
@aarongarciah This regression test fails because it did not render `Typography` since the beginning, but typographic styles are now applied via `theme.typography.body1`. The issue is that keeping the `Typography` component in the rest of the demos results in invalid HTML because a `<p>` (Typography's default element) cannot be inside a `<span>`. Removing `Typography` causes significant visual regressions. To address this, I applied `theme.typography.body1` directly to the Accordion Summary.

Options:
1. Remove `theme.typography.body1` from accordion summary and `Typography` from demos, accept the visual regressions, and leave styling to the developer without enforcing typographic `body1` style.
2. Remove `theme.typography.body1` and keep `Typography` in the demos but use `<Typography component="span" />` for valid HTML. However, this feels like a workaround.

Apologies for not flagging this earlier.

> I just added the `v7.x` label

I think this should be released in v6. This can be considered a fix. The breaking change was already done in #42914 with the introduction of heading. We shouldn't have a breaking change which is wrong. What do you think?

aarongarciah commented:
@ZeeshanTamboli option 1 sounds like a breaking change, so I think this option shouldn't be introduced in v6. Option 2 sounds safer. Why using `<Typography component="span" />` feels like a workaround? Changing the HTML tag is something `Typography` support so I'm curious why you feel that way.

cc @DiegoAndai in case you have some thoughts.

ZeeshanTamboli commented:
> Why using <Typography component="span" /> feels like a workaround?

@aarongarciah Using `<Typography component="span" />` initially felt like a workaround because it adds an extra element just to ensure valid HTML. However, after reconsidering, it doesn’t seem like a workaround anymore.  

So I tried option 2: removing `theme.typography.body1` and keeping `Typography` in the demos with `<Typography component="span" />` for valid HTML.  

But the issue is that removing `theme.typography.body1` causes default HTML button styles to apply (see screenshot below). This forces the user to always use a Typography component for styling. IMO, the current approach what we have right now is better. 

![Screenshot (34)](https://github.com/user-attachments/assets/2d86e15c-b960-4c44-add8-9c485998ef34)

DiegoAndai commented:
I agree with @aarongarciah that the changes proposed are breaking and cannot be released in `6.x`, and we should try to find a workaround in the meantime

> But the issue is that removing theme.typography.body1 causes default HTML button styles to apply (see screenshot below). This forces the user to always use a Typography component for styling the summary text.

What about removing the button styles?

ZeeshanTamboli commented:
> What about removing the button styles?

How to do that?

DiegoAndai commented:
I'm a bit lost.

This is the option I think we should go for:

> Option 2: Remove theme.typography.body1 and keep Typography in the demos but use <Typography component="span" /> for valid HTML. However, this feels like a workaround.

Why, then, are the typography styles not applied if we're still using the `Typography` component?

ZeeshanTamboli commented:
> I'm a bit lost.
> 
> This is the option I think we should go for:
> 
> > Option 2: Remove theme.typography.body1 and keep Typography in the demos but use  for valid HTML. However, this feels like a workaround.
> 
> Why, then, are the typography styles not applied if we're still using the `Typography` component?

@DiegoAndai See this sentence: This forces the user to always use a Typography component for styling the summary text.

DiegoAndai commented:
Maybe we could:

- Change the Accordion summary element to `button`.
- Update the Accordion summary content and icon wrapper elements to `span`.
- Keep `Typography` on the demos as before, but with `component='span'`

And release that as a minor.

ZeeshanTamboli commented:
I've made the changes. The first demo no longer uses Typography, so Argos CI flagged the change (aside from the flaky progress component test). I wasn't aware the first demo differed between CodeSandbox and the docs, but now they're the same. The docs theme isn't applied—is that okay? I think it is because the HTML button styles are applied, as shown in the screenshot:

![Screenshot (38)](https://github.com/user-attachments/assets/5b22a178-5871-416b-8136-1559c8723fa1)


**This PR:** https://deploy-preview-44408--material-ui.netlify.app/material-ui/react-accordion/#introduction

**Production:** https://mui.com/material-ui/react-accordion/#introduction)](https://mui.com/material-ui/react-accordion/#introduction

aarongarciah commented:
@ZeeshanTamboli can you update the first demo to include the `Typography` component so it looks the same as the other demos?

ZeeshanTamboli commented:
> @ZeeshanTamboli can you update the first demo to include the `Typography` component so it looks the same as the other demos?

@aarongarciah Updated.

aarongarciah reviewed (commented):
Looks good! Thanks for your patience.

@ZeeshanTamboli I think only 2 things are missing:
- Add an entry in the migration guide explaining the change.
- Update the PR description to reflect the latest changes in the PR.

In the next release, we'll explain the change, too.

ZeeshanTamboli commented:
@aarongarciah  
> * Add an entry in the migration guide explaining the change.

Added. https://deploy-preview-44408--material-ui.netlify.app/material-ui/migration/upgrade-to-v6/#accordion-starting-from-v6-3-0

> * Update the PR description to reflect the latest changes in the PR.

Updated

aarongarciah commented on docs/data/material/migration/upgrade-to-v6/upgrade-to-v6.md:
````diff
@@ -155,6 +155,13 @@ If the default heading element conflicts with your existing structure, you can c
 </Accordion>
 ```
 
+### Accordion (Starting from v6.3.0)
````
I think it'd be better to merge this section with the previous one. Maybe something like this (with better copy? 😅):

```
### Accordion

#### Summary wrapped in a heading

#### Summary as a button (from v6.3.0)
```

ZeeshanTamboli replied:
Makes sense. Done.

DiegoAndai reviewed (approved):
LGTM ✅ 

As always, @ZeeshanTamboli, thanks for working on this and willingness to discuss approaches to get the best solution for the community.

Let's wait for a final review from @aarongarciah.

DiegoAndai commented:
CC: @mnajdova for next week's release: this required a minor bump and a mention on the CHANGELOG.

aarongarciah reviewed (approved):
Nice!

DiegoAndai commented:
I argue this can be considered as a fix (of previously invalid HTML) more than a breaking change so I'm removing the "breaking" label and instead labeling as "minor version change".

Paper-Folding commented:
By changing the element to `button`, use of `IconButton`(as a helper button) inside `AccordionSummary` now raise an hydration error, although the error does not causing any damage, how can I get away it?

ZeeshanTamboli commented:
> By changing the element to `button`, use of `IconButton`(act as a helper button in our project) inside `AccordionSummary` now raise a hydration error, although the error does not causing any damage, how to get away it?

@Paper-Folding Looks like your issue is similar to #45187

Paper-Folding commented:
> > By changing the element to `button`, use of `IconButton`(act as a helper button in our project) inside `AccordionSummary` now raise a hydration error, although the error does not causing any damage, how to get away it?
> 
> @Paper-Folding Looks like your issue is similar to #45187

Yeah, react complains about that "In HTML, <button> cannot be a descendant of <button>.", so, why mui have to follow so-called "W3C Accordion Pattern Standards"? :)

ZeeshanTamboli commented:
> > > By changing the element to `button`, use of `IconButton`(act as a helper button in our project) inside `AccordionSummary` now raise a hydration error, although the error does not causing any damage, how to get away it?
> > 
> > 
> > @Paper-Folding Looks like your issue is similar to #45187
> 
> Yeah, react complains about that "In HTML, <button> cannot be a descendant of <button>.", so, why mui have to follow so-called "W3C Accordion Pattern Standards"? Even using a `<section>` is better than that. :)

All components are ARIA-compliant, but this was a bug where it wasn’t before ARIA-compliant. So we had to change the HTML structure. 
Avoid nesting buttons inside buttons—use an adjacent button next to the accordion summary header instead. This is also mentioned in the [ARIA docs](https://www.w3.org/WAI/ARIA/apg/patterns/accordion/):
> In some accordions, there are additional elements that are always visible adjacent to the accordion header. For instance, a menubutton may accompany each accordion header to provide access to actions that apply to that section
